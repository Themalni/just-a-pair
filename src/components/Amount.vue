<template>
  <div>
    <span class="font-weight-bold d-block mt-2 mb-2">{{ label }}</span>
    <div class="amount">
      <button
          class="btn m-0"
          :disabled="selectedAmount === 1"
          @click="decreaseAmount"
      >
          -
      </button>
      <span class="amount-number border-dark">{{ selectedAmount }}</span>
      <button class="btn m-0" @click="increaseAmount">+</button>
    </div>
  </div>
  </template>

<script>

import { mapState } from "vuex";

export default {
  name: "Amount",
  props: {
    label: {
        type: String,
        default: "Select amount"
    }
  },

  data() {
    return {
        selectedAmount: 1
    }
  },

  watch: {
    productPreviewId() {
      this.resetAmount();
    }
  },

  computed: {
    ...mapState(["productPreview"]),

    productPreviewId() {
      return this.productPreview.id;
    }
  },

  methods: {
    resetAmount() {
      this.selectedAmount = 1;
    },
    increaseAmount() {
      this.selectedAmount = ++this.selectedAmount;
      this.$emit('amount', this.selectedAmount);
    },
    decreaseAmount() {
      if(this.selectedAmount <= 1) {
          return;
      }
      this.selectedAmount = --this.selectedAmount;
      this.$emit('amount', this.selectedAmount);
    }
  }
}
</script>

<style lang="scss" scoped>
.amount {
  display: inline-flex;
  align-items: center;

  .btn {
    padding: 5px 10px;
    box-shadow: none;
    color: inherit;
    border: 2px solid $dark-grey;
    line-height: 16px;

    &:first-of-type {
      border-right: none;
      border-top-right-radius: 0;
      border-bottom-right-radius: 0;
    }
    &:last-of-type {
      border-left: none;
      border-top-left-radius: 0;
      border-bottom-left-radius: 0;
    }

    &:disabled {
      border-color: $grey;
      color: $grey;
    }
  }
}
.amount-number {
    border: 2px solid $dark-grey;
    padding: 5px 10px;
    line-height: 16px;
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
}
</style>