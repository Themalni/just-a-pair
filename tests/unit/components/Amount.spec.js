import { mount, createLocalVue } from '@vue/test-utils';
import Amount from '@/components/Amount';
import mockState from '../utils/mockState';
import Vuex from 'vuex';


const localVue = createLocalVue();
localVue.use(Vuex);

describe('Amount', () => {
    let wrapper
    let store
    let state
    let spy

    beforeEach(() => {
        // state: mockState
        store = new Vuex.Store({
            state: mockState
        })

        wrapper = mount(Amount, {
            store,
            localVue,
            propsData: {
                label: 'Select product amount'
            }
        });
    });

    afterEach(() => {
        wrapper.destroy();
    })

    it('renders correctrly', () => {
        expect(wrapper.html()).toMatchSnapshot();
    })

    it('renders label correctly', () => {
        expect(wrapper.vm.label).toBe('Select product amount')
    })

    it('resets amount to default value', async () => {
        await wrapper.setData({ selectedAmount: 3 })
        expect(wrapper.vm.selectedAmount).toBe(3)

        wrapper.vm.resetAmount();
        expect(wrapper.vm.selectedAmount).toBe(1)
    })

    it('increases amount by `1`', async () => {
        wrapper.vm.increaseAmount();
        expect(wrapper.vm.selectedAmount).toBe(2)
    })

    it('emits increased value', async () => {
        const selectedAmount = wrapper.vm.selectedAmount;

        wrapper.vm.increaseAmount();
        wrapper.vm.$emit('amount', selectedAmount);

        expect(wrapper.emitted().amount).toBeTruthy();
        expect(wrapper.emitted().amount[0]).toEqual([2])
    })

    it('decreases amount by `1`', async () => {
        await wrapper.setData({ selectedAmount: 3 })
        wrapper.vm.decreaseAmount();
        expect(wrapper.vm.selectedAmount).toBe(2)
    })

    it('decrease amount is not called if amount is less or qual to `1`', () => {
        wrapper.vm.decreaseAmount();
        expect(wrapper.vm.selectedAmount).toBe(1);
    })

    it('emits deccreased value', async () => {
        await wrapper.setData({ selectedAmount: 3 })
        const selectedAmount = wrapper.vm.selectedAmount;

        wrapper.vm.decreaseAmount();
        wrapper.vm.$emit('amount', selectedAmount);

        expect(wrapper.emitted().amount).toBeTruthy();
        expect(wrapper.emitted().amount[0]).toEqual([2])
    })

    it('resets selected amount when product id has been changed', async () => {
        console.log("productPreview:", wrapper.vm.$store.state)
        // wrapper.vm.$store.state.productPreview.id = 999;
        // wrapper.vm.$nextTick(() => {
        //     expect(wrapper.vm.resetAmount()).toBeCalled();
        //     done();
        // });
        // wrapper.vm.$options.watch.$store.state.productPreview.id.call(wrapper.vm, 999);
        // state.productPreview.id = 622;
        
    })
})