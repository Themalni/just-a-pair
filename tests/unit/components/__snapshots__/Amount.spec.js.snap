// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Amount renders correctrly 1`] = `
<div><span class="font-weight-bold d-block mt-2 mb-2">Select product amount</span>
  <div class="amount"><button disabled="disabled" class="btn m-0">
      -
    </button> <span class="amount-number border-dark">1</span> <button class="btn m-0">+</button></div>
</div>
`;
